### PR TITLE
RefCounted implementation for classes

### PR DESCRIPTION
**Pull request by MartinNowak**
- add a simple RefCounted variant for classes
- varies quite a lot from the RefCounted for structs
  because certain semantics aren't possible with classes
- also remove the autoInit stuff and use a function+@disable
  this() to guarantee that the payload is initialized

### CODE REVIEW

**Comment by MartinNowak** [comment]
This still needs a bit polishing (docs et.al.), but I wanted to discuss a few things upfront.
The implementation for classes varies quite a lot from the existing RefCounted (which is also a PITA due to the autoInit stuff), but as things like autoInit and opAssign aren't implementable for classes and most of the other functions differ, I just made a second RefCounted.

**Comment by rsw0x** [comment]
Could you add an @nogc unit test(AFAICT this can infer @nogc, right?)

**Comment by JakobOvrum** [comment]
I agree this doesn't need the lazy initialization stuff, but instead of making it non-nullable, why not make it work just like any class reference `T` and be `null` by default? It should make it easier to use, usable in generic code that needs default-initialization, and the user should be able to make it non-null by composing it with a `NotNull` type.

It shouldn't be hard to implement; just cast `_store` instead of `_store.payload` (although either would probably work even when `_store` is `null`) and then add the proper overloads for `typeof(null)`.

**Comment by MartinNowak** [comment]
Added the @nogc test, adjusted the methods of the old RefCounted, and enabled ref escape checking for that.

> why not make it work just like any class reference T and be null by default

Because it's a class value, not a class reference, i.e. you can't nullify `scoped!Class` either.
Granted it's possible to special case `null`, but being able to do `rc = null;` seems weird when you can't do `rc = cinst;`.

TODO:

- [ ] implement polymorphic RefCounted conversions

**Comment by MartinNowak** [comment]
In fact using payload and refCount might hide members of the payload, will change that back to the old scheme of wrapping everything in refCountedStore.

**Comment by JakobOvrum** [comment]
> Because it's a class value, not a class reference, i.e. you can't nullify `scoped!Class` either.

`RefCounted!T` is a reference by any account I can think of. It is implemented in terms of a single pointer, and copying it makes an alias, not a copy of the payload. `Ref` is right there in the name.

> Granted it's possible to special case `null`, but being able to do `rc = null;` seems weird when you can't do `rc = cinst;`.

You can if `cinst` is also a `RefCounted`.

> How would you initialize a null RefCounted!T?

Analogous to `T`:
```d
// These are equivalent
RefCounted!T c;
RefCounted!T c2 = null;
```

**Comment by MartinNowak** [comment]
Mmh, I guess we could add a null state for RefCounted.
Default construction has problems with a nullary class constructors though.
```d
auto c = RefCounted!C(); // this is null
auto c2 = refCounted!C(); // this is not
```

**Comment by JakobOvrum** [comment]
Ouch. `std::shared_ptr` gets around this by delegating the onus of construction to the user, but this is not an option for us.

**Comment by MartinNowak** [comment]
This difference might be OK though, because one is clearly a function call while the other is a struct literal.

File: std/typecons.d [hunk 1]
```diff
@@ -5000,6 +5088,83 @@ unittest
...
+
+    assert(Foo.ctor == 0 && Foo.dtor == 0);
+    {
+        auto rc = refCounted!Foo();
+        assert(Foo.ctor == 1 && Foo.dtor == 0);
+        auto rc2 = rc;
+        assert(Foo.ctor == 1 && Foo.dtor == 0);
+    }
+    assert(Foo.ctor == 1 && Foo.dtor == 1);
+}
```
Comment (burner) [inline std/typecons.d]: could add a test for the inheritance case

**Comment by MartinNowak** [comment]
Changed the implementation to account for the discussed changes.
There is still quite some stuff left to be done (see TODOs), most importantly attribute inference doesn't work correctly.
I'd highly appreciate any help in getting this done, e.g. improving the documentation and examples or implementing nullable for Unique.

File: std/typecons.d [hunk 2]
```diff
@@ -4865,12 +4865,154 @@ unittest
...
+
+    /// Returns storage implementation struct.
+    @property nothrow @safe
+    ref inout(RefCountedStore) refCountedStore() inout
+    {
+        return _refCounted;
+    }
+
+    /// Boolean conversion for RefCounted, true when initialized.
+    bool opCast(T:bool)() const @safe
```
Comment (JakobOvrum) [inline std/typecons.d]: This notation catches things like `cast(int)` as well, it's probably better to use `bool opCast(T)() if(is(T == bool))`

File: std/typecons.d [hunk 3]
```diff
@@ -4865,12 +4865,154 @@ unittest
...
     auto rc2 = rc1;
+    assert(rc1.refCountedStore.refCount == 2);
     // Reference semantics
     rc2 = 42;
     assert(rc1 == 42);
     // the pair will be freed when rc1 and rc2 go out of scope
 }
 
+// recursively destroy a class instance
+private void _destroy(T)(T obj) if (is(T == class))
```
Comment (JakobOvrum) [inline std/typecons.d]: This doesn't handle destruction of fields - is there any reason object.destroy can't be used here?

File: std/typecons.d [hunk 4]
```diff
@@ -4865,12 +4865,154 @@ unittest
...
+     }
+     // leak monitor b/c destruction would be impure, will be collected by the GC
+}
+
+/// RefCounted implementation for classes.
+struct RefCounted(T)
+    if (is(T == class))
+{
+    /// $(D RefCounted) storage implementation.
+    struct RefCountedStore
```
Comment (JakobOvrum) [inline std/typecons.d]: Is there any benefit to exposing this?

File: std/typecons.d [hunk 5]
```diff
@@ -4865,12 +4865,154 @@ unittest
...
+    }
+
+    /// Boolean conversion for RefCounted, true when initialized.
+    bool opCast(T:bool)() const @safe
+    {
+        return _refCounted.isInitialized;
+    }
+
+    /// copy constructor for other RefCounted class
+    this(U)(RefCounted!U u)
```
Comment (JakobOvrum) [inline std/typecons.d]: Not a huge deal, but with `void opAssign(typeof(null))` there should probably also be a corresponding `this(typeof(null))`.

File: std/typecons.d [hunk 6]
```diff
@@ -4865,12 +4865,154 @@ unittest
...
+        _refCounted._store = null;
+    }
+
+    /**
+       Get the underlying payload.
+
+       Note that getting a class reference is currently unsafe
+       as there is currently no way to stop it from escaping. (see DIP69)
+     */
+    @property nothrow @system
```
Comment (JakobOvrum) [inline std/typecons.d]: Any reason it's explicitly marked `nothrow` but relies on attribute inference for `pure` and `@nogc`?

File: std/typecons.d (hunk 5 above)
Comment (MartinNowak) [inline std/typecons.d]: The default constructor all already be null initialized, but for symmetry it makes sense to add it.

File: std/typecons.d (hunk 4 above)
Comment (MartinNowak) [inline std/typecons.d]: The existing RefCounted namespaces everything under the struct which needs to be documented for functions like RefCounted.

File: std/typecons.d (hunk 3 above)
Comment (MartinNowak) [inline std/typecons.d]: Lack of attribute inference when using rt_finalize.

File: std/typecons.d (hunk 2 above)
Comment (MartinNowak) [inline std/typecons.d]: No it does not `Error: template instance opCast!int does not match template declaration opCast(T : bool)()`, parameter specializations are exact matches despite the colon.

Comment (JakobOvrum) [inline std/typecons.d]: Huh, didn't know that, thanks.

File: std/typecons.d (hunk 6 above)
Comment (MartinNowak) [inline std/typecons.d]: Nope

File: std/typecons.d (hunk 3 above)
Comment (JakobOvrum) [inline std/typecons.d]: Eek, I thought I dealt with all attribute inference cases :(

I haven't looked into rt_finalize, but as I understand it, it does destruction as well as zero-bombing the whole class instance, crucially including the vptr. Does it do anything else? Otherwise, it could be implemented generically like `destructRecurse` did for structs and fixed-length arrays.

File: std/typecons.d (hunk 2 above)
Comment (MartinNowak) [inline std/typecons.d]: It's weird that the syntax deviates from the is expression, just another accidental complexity.

**Comment by MartinNowak** [comment]
Left to be done.
- [ ] implement proper destruction
- [ ] polish documentation

File: std/typecons.d (hunk 4 above)
Comment (JakobOvrum) [inline std/typecons.d]: Is there anything that should be user-facing apart from `refCount`? If not, it could just be moved to `RefCounted` itself. There is a cost to public symbols because of the generic AliasThis, but it already has `refCountedStore` for this, so it would just remove one and add one.

File: std/typecons.d (hunk 5 above)
Comment (JakobOvrum) [inline std/typecons.d]: Some people, myself included, like to use explicit null-initialization in certain situations where it helps readability :)

File: std/typecons.d (hunk 3 above)
Comment (MartinNowak) [inline std/typecons.d]: And we still need to replace `destructRecurse` with a compiler based implementation :).

Comment (JakobOvrum) [inline std/typecons.d]: Or amend the compiler to use `destructRecurse` :)

Comment (MartinNowak) [inline std/typecons.d]: That's super complex because one would need to interleave field destructors with base class destructors.

File: std/typecons.d (hunk 4 above)
Comment (MartinNowak) [inline std/typecons.d]: It's a bit ugly to have a different RefCounted API based on whether it's a class or a struct. There is also `isInitialized`.

Comment (JakobOvrum) [inline std/typecons.d]: Well, we overhauled `Unique` because it was plain bad, if it turns out we can find a better API here, we should do it.

I think we should stick as close to C++'s smart pointers as possible, but existing `Unique` and `RefCounted` both make that difficult. Their idiosyncratic behaviours make both smart pointers unsuitable for generic code and clumsy in any kind of code. This is mostly because of the inconsistent and/or poorly implemented null state, but I don't see a reason why `RefCounted` should expose the internal type used to put the value on the heap just to expose a couple of pieces of metadata. `isInitialized` wouldn't have existed in the first place if `RefCounted` was implemented as a proper smart pointer with a consistently expressed null state... the reference count is truly the only piece of metadata and thus the only thing that warrants a named member function in `RefCounted`.

I really don't know what the deal is with the whole auto-initialization thing. Lazily initialized objects tend to suck in practice, causing bugs where an intent to alias some object actually didn't because the reference was null at that time, despite there never being an explicit initialization anywhere to be found in the entire codebase. I've had issues with AAs behaving like this. Lazy initialization does *not* make reference types behave like value types.

/rant

**Comment by mrkline** [comment]
I am terribly sorry I didn't chime in sooner - I was on vacation the past two weeks.

A lot of what @JakobOvrum has said above really resonates with my. Frankly, if we're going to overhaul `Unique` with this upcoming release, `RefCounted` should get a similar makeover. The two go hand in hand. As we all discussed in the `Unique` PR, now is not the time for half-measures and just putting a coat of polish on the old versions. Now is the a good time to rip them up and Do It Right™. `RefCounted` should:

- Have much in common with the new `Unique`, including
  - `get()` functions
  - an `empty` property and `opCast(T:bool)` indicating whether the underlying pointer is null
  - `alias get this`
  - A single `refCounted` function which takes the constructor arguments for the underlying type and returns a `RefCounted`

- Get rid of all the lazy initialization stuff such as `autoInit` and `ensureInitialized`. This is a can of worms a smart pointer type shouldn't open.

- Get rid of `refCountedStore()`. The user shouldn't concern themselves with the underlying storage. That's an  implementation detail the user should have no need to know, provided we provide correct reference counting semantics.

In short, its API should be nearly identical to that of `Unique` (possibly with the addition of a `refCount` property) , just like `shared_ptr` has so much in common with `unique_ptr`.

Again, I'm sorry for all the legwork you've already done, Martin - I'll help in any way I can now that I'm back, including taking up these changes in my own PR.

**Comment by MartinNowak** [comment]
> A lot of what @JakobOvrum has said above really resonates with my. Frankly, if we're going to overhaul Unique with this upcoming release, RefCounted should get a similar makeover. The two go hand in hand. As we all discussed in the Unique PR, now is not the time for half-measures. Now is a good time to rip them up and Do It Right™. RefCounted should:

Not quite, `RefCounted` is widely used, so we cannot make radical changes.

> get() functions
> an empty property and opCast(T:bool) indicating whether the underlying pointer is null
> alias get this

The implicit bool conversion is there, but `get` and `empty` could hide functions of the underlying type making things like a RefCounted range impossible. We should probably revise `Unique` to account for this.

> A single refCounted function which takes the constructor arguments for the underlying type and returns a RefCounted

Will extend `refCounted` it to non-class types.

> Get rid of all the lazy initialization stuff such as autoInit and ensureInitialized(). This is a can of worms a smart pointer type shouldn't open.

I agree, we'll have to see how to deprecate it properly.

> Get rid of the refCountedStore() function. The user shouldn't concern themselves with the underlying storage. That's an implementation detail the user should have no need to know, provided we provide correct reference counting semantics.

I did that before, named it `payload`, but the purpose of refCountedStore is to namespace functions of RefCounted. Might make sense to drop this, b/c most RefCounted methods are unambiguous.

> In short, its API should be nearly identical to that of Unique (possibly with the addition of a refCount property) , just like shared_ptr has so much in common with unique_ptr.

Right, let's see how far we can get.

> Again, I'm sorry for all the legwork you've already done, Martin - I'll help in any way I can now that I'm back, including taking up these changes in my own PR.

You can make PRs that target my [branch](https://github.com/MartinNowak/phobos/tree/master) to collaborate on this.

**Comment by MartinNowak** [comment]
> The implicit bool conversion is there, but get and empty could hide functions of the underlying type

I think `get` and `refCount` should be fine, `empty` isn't needed as `opCast(T:bool)()` can be used instead.

**Comment by mrkline** [comment]
Ah, I didn't realize that `RefCounted` was more widely used, so perhaps we need to move a bit more slowly. I also forgot Github lets you target PRs at arbitrary branches. I'll get poking around. Thanks again for the help.

**Comment by MartinNowak** [comment]
I added support for any type in `refCounted`.

> Ah, I didn't realize that RefCounted was more widely used, so perhaps we need to move a bit more slowly. I also forgot Github lets you target PRs at arbitrary branches. I'll get poking around. Thanks again for the help.

Meanwhile I'll work on the destruction problem.

**Comment by JakobOvrum** [comment]
> The implicit bool conversion is there, but `get` and `empty` could hide functions of the underlying type making things like a RefCounted range impossible. We should probably revise `Unique` to account for this.

I have proposed to remove `Unique.empty` in #3225.

**Comment by mrkline** [comment]
I'll hit the ground with this once after we settle on what API we want in #3225.

**Comment by MartinNowak** [comment]
> Meanwhile I'll work on the destruction problem.

Done, https://github.com/D-Programming-Language/dmd/pull/4650.

**Comment by rsw0x** [comment]
ping, what's the status of this?

**Comment by mrkline** [comment]
I believe we're waiting for #3225 to shake out before deciding what to do here.

**Comment by MartinNowak** [comment]
TODO:

- [ ] Use isNull and nullify API
- [ ] use __xdtor to recursively destruct classes

**Comment by MartinNowak** [comment]
Closing in favor of a complete PR of all smart refs (WIP https://github.com/MartinNowak/phobos/tree/smartRefs).
